### PR TITLE
Make the config versioning/compatibility independent of the butido version and improve the config checks

### PR DESCRIPTION
**Pull request by primeos-work**
tl;dr:
- Version the butido configuration with integer numbers (vs. SemVer version requirements)
- Check if the config is compatible before trying to load (type check) it to improve the error messages
- Automatically output the required configuration changes (changelog) if updates are required
- Add tests for the example config files in this repo
- Fail with an error if there are unknown settings in the configuration
- Improve the checks for missing directories (one missing check + two wrong config key names)

Note: This makes breaking configuration changes that require users to update their configuration (should be trivial though as one only has to set `compatibility=1` and remove invalid settings, if present).

<details>
<summary>Example outputs</summary>

Before this PR:
```
$ butido build -I rh8 tree 1.8.0
Error: Failed to validate configuration

Caused by:
    Configuration is not compatible to butido 0.4.0
```

If config upgrades are required:
```
$ butido build -I rh8 tree 1.8.0
The butido configuration is too old and the following changes are required:
- Version 1: The format of the `compatibility` setting has changed from a string (`semver::VersionReq`) to a number (`u16`).
- Update the `compatibility` setting to `1`

Error: The butido configuration failed the compatibility check

Caused by:
    0: The expected configuration version is 1 while the provided configuration has a compatibility setting of 0
    1: The provided configuration is not compatible with this butido binary
```

Trying to load an old configuration:
```
$ butido build -I rh8 tree 1.8.0
Error: The butido configuration failed the compatibility check

Caused by:
    0: Set "compatibility" to 0 to get a summary of the required changes
    1: The format of the "compatibility" setting has changed from a string to a number
    2: Failed to parse the value of the compatibility setting (0.4.0) into a number (str -> u16)
    3: invalid digit found in string
```

Loading a configuration with invalid settings:
```
$ butido build -I rh8 tree 1.8.0
Error: Failed to load (type check) the butido configuration

Caused by:
    unknown field `invalid`, expected one of `compatibility`, `log_dir`, `strict_script_interpolation`, `progress_format`, `spinner_format`, `package_print_format`, `build_error_lines`, `script_highlight_theme`, `script_linter`, `shebang`, `releases_root`, `release_stores`, `staging`, `source_cache`, `database_host`, `database_port`, `database_user`, `database_password`, `database_name`, `database_connection_timeout`, `docker`, `containers`, `available_phases`
```

Invalid env vars:
```
$ BUTIDO_TEST=42 butido build -I rh8 tree 1.8.0
Error: Failed to load (type check) the butido configuration

Caused by:
    unknown field `test`, expected one of `compatibility`, `log_dir`, `strict_script_interpolation`, `progress_format`, `spinner_format`, `package_print_format`, `build_error_lines`, `script_highlight_theme`, `script_linter`, `shebang`, `releases_root`, `release_stores`, `staging`, `source_cache`, `database_host`, `database_port`, `database_user`, `database_password`, `database_name`, `database_connection_timeout`, `docker`, `containers`, `available_phases`
```
</details>

### CODE REVIEW

**Review by christophprokop — APPROVED** [review]
Tested. Thank you!
